### ISSUE TITLE
[Runtime Module] Quadratic Coinvoting Democracy

### ISSUE THREAD

**Issue by shawntabrizi** [substrate-developer-hub/hacktoberfest#22]
This runtime module would allow users of a Substrate blockchain to vote on proposals using their tokens in a quadratic voting system.

The goal of this module is to provide alternative governance mechanisms to what the SRML provides. Right now, Substrate has a "lock weighted" voting mechanism, but it is worth considering a quadratic coin voting system.

To do this task, it would require you do a bit of research on quadratic voting, and understand if it makes sense in the context of Substrate/blockchains. There may be different implementation details between papers, so you will have to pick one which makes sense to you.

Ideally, this module will expose the same traits as the existing Democracy module, thus allowing it to simply be swapped into any Substrate runtime which currently uses the SRML Democracy module.

However, I think any working proof of concept may be a good starting point for the context of "hacktoberfest".

For mentoring this issue please contact: [me, @shawntabrizi](https://substrate.dev/hacktoberfest/hackers/shawntabrizi.html)

## Proof of done

 - [ ] Create a new project by copying the [contents from our runtime module template](https://github.com/substrate-developer-hub/substrate-module-template)
 - [ ] Don't forget to add an OpenSource License !
 - [ ] Implement the specific module
 - [ ] Once done, submit a PR either of the module in a subfolder under `submissions` or as a git submodule under `subfolder`
 - [ ] (optionally) publish if on crates.io

**Probable/Expected Workload**:
 - [ ] minimal (< 1h)
 - [ ] easy (a few hours)
 - [ ] intermediate (4-8h)
 - [ x ] substantial (1-3 days)
 - [ ] involved (up to a week)
 - [ ] intense (more than a week)
 - [ ] impossible (more than a month)

**Comment by filippetroviccc** [comment]
Hey @shawntabrizi ,

Is this issue still open and up for grabs even doe ` Hacktoberfest 2019` was finished?

**Comment by shawntabrizi** [comment]
Sure! If you can get an MVP started, and are interested in taking it further, you might possibly consider applying for a grant!

https://web3.foundation/grants/

**Comment by filippetroviccc** [comment]
Excellent, I will start with MVP first and then I will ping you to talk a bit more about what features does it need to have in order to be an actual runtime module.

**Comment by filippetroviccc** [comment]
Hey @shawntabrizi ,

I was researching and thinking a bit about the approach that we should take with the quadratic module. I think we could extend the current democracy pallet with a new code and with some new functionality. If you want we can get on a call or should I write an idea here (it would probably be little too much of text).

**Comment by shawntabrizi** [comment]
The quadratic coin voting should be a separate module that could replace the existing democracy pallet (i.e. provides all the same Rust traits).

Unfortunately I am not really free to mentor at the moment, but maybe after we launch polkadot. But really the better path forward would be just to try some stuff and request feedback.

**Comment by filippetroviccc** [comment]
Cool, Here is just a brief explanation about the approach that we are going to take:

The current democracy pallet has fixed vote weight that is pegged to Balance. The idea is to have dynamic VoteWeight (similar to VoteThreshold) were the proposer of the referendum will choose what type of weight he would like to use (e.g. Standard, Quadratic, etc.). On `try_vote` we will just calculate what is the actual weight of the vote with a formula that will be implemented for the desired approach. This could either be a separate module or part of democracy (it is a case of removing VoteWeight::Standard()) and other people would be able to add a new vote weight algorithm pretty easy.

I'll ping you for some feedback when we finish. Best.

**Comment by filippetroviccc** [comment]
Hey @shawntabrizi ,

I have finished implementing a quadratic-democracy pallet in my own substrate fork. Where should I submit PR? This repo is kinda old and possible there is better and faster way for me to get feedback.

**Comment by shawntabrizi** [comment]
I dont think we would be looking to accept PRs into Substrate directly, but ideally you might be able to contribute it to the Substrate Developer Hub.

Why not share a link here and we can take a look

**Comment by filippetroviccc** [comment]
Here is the link: https://github.com/MVPWorkshop/substrate-quadratic-democracy

Also to point out that `/bin/node` is switched to work with `quadratic_democracy` instead of normal democracy.

**Comment by filippetroviccc** [comment]
@shawntabrizi the link is above.

**Comment by shawntabrizi** [comment]
@filippetroviccc I dont know when I will have a chance to look at this. I have some of my own work taking my time at the moment.

Maybe it would help to get context of what your end goal is here? Just to learn? You want to merge this into Substrate?

**Comment by filippetroviccc** [comment]
Thank you Shawn. Sorry for pinging you again, I know you are really busy and this for sure is not on top of your priority list (probably at the bottom :) ). I would like to get feedback on the code, and no need to merge it in Substrate I can open PR in Substrate Developer Hub. I was just wondering if I should wait for your feedback or can I open PR immediately.
